filters/auth/grant.go

```diff
@@ -0,0 +1,197 @@
+}
+
+func loginRedirectWithOverride(ctx filters.FilterContext, config OAuthConfig, originalOverride string) {
+	req := ctx.Request()
+	redirect, original := config.RedirectURLs(req)
+
+	if originalOverride != "" {
+		original = originalOverride
+	}
+
+	state, err := config.flowState.createState(original)
+	if err != nil {
+		log.Errorf("Failed to create login redirect: %v", err)
+		serverError(ctx)
+		return
+	}
+
+	authConfig := config.GetConfig()
+	ctx.Serve(&http.Response{
+		StatusCode: http.StatusTemporaryRedirect,
+		Header: http.Header{
+			"Location": []string{authConfig.AuthCodeURL(state, config.GetAuthURLParameters(redirect)...)},
+		},
+	})
+}
+
+func (f *grantFilter) refreshToken(c cookie) (*oauth2.Token, error) {
+	// Set the expiry of the token to the past to trigger oauth2.TokenSource
+	// to refresh the access token.
+	token := &oauth2.Token{
+		AccessToken:  c.AccessToken,
+		RefreshToken: c.RefreshToken,
+		Expiry:       time.Now().Add(-time.Minute),
+	}
+
+	ctx := providerContext(f.config)
+
+	// oauth2.TokenSource implements the refresh functionality,
+	// we're hijacking it here.
+	tokenSource := f.config.GetConfig().TokenSource(ctx, token)
+	return tokenSource.Token()
+}
+
+func (f *grantFilter) refreshTokenIfRequired(c cookie) (*oauth2.Token, error) {
+	now := time.Now()
+	isAccessTokenExpired := c.isAccessTokenExpired()
+	canRefresh := !f.config.DisableRefresh && c.RefreshToken != ""
+	shouldRefresh := isAccessTokenExpired || now.After(c.RefreshAfter)
+
+	if shouldRefresh {
+		if canRefresh {
+			return f.refreshToken(c)
+		} else {
+			return nil, errExpiredToken
+		}
+	} else {
+		return &oauth2.Token{
+			AccessToken:  c.AccessToken,
+			TokenType:    "Bearer",
+			RefreshToken: c.RefreshToken,
+			Expiry:       c.Expiry,
+		}, nil
+	}
+}
+
+func (f *grantFilter) setAccessTokenHeader(req *http.Request, token string) {
+	if f.config.AccessTokenHeaderName != "" {
+		req.Header.Set(f.config.AccessTokenHeaderName, authHeaderPrefix+token)
+	}
+}
+
+func (f *grantFilter) createTokenContainer(token *oauth2.Token, tokeninfo map[string]interface{}) tokenContainer {
+	subject := ""
+	if f.config.TokeninfoSubjectKey != "" {
+		subject = tokeninfo[f.config.TokeninfoSubjectKey].(string)
+	}
+
+	tokeninfo["sub"] = subject
+
+	return tokenContainer{
+		OAuth2Token: token,
+		Subject:     subject,
+		Claims:      tokeninfo,
+	}
+}
+
+func (f *grantFilter) Request(ctx filters.FilterContext) {
+	req := ctx.Request()
+
+	c, err := extractCookie(req, f.config)
+	if err == http.ErrNoCookie {
+		loginRedirect(ctx, f.config)
+		return
+	}
+
+	token, err := f.refreshTokenIfRequired(*c)
+	if err != nil && c.isAccessTokenExpired() {
+		// Refresh failed and we no longer have a valid access token.
+		loginRedirect(ctx, f.config)
+		return
+	}
+
+	tokeninfo, err := f.config.TokeninfoClient.getTokeninfo(token.AccessToken, ctx)
+	if err != nil {
+		if err == errInvalidToken {
+			log.Errorf("Error while calling tokeninfo: %v.", err)
```

Comment:
Error while calling -> Failed to call

Reply:
I think the solution is to change `==` to `!=`. We only need to log the unknown errors when calling tokeninfo.

Reply:
or if not c.isAccessTokenExpired()

Reply:
i think the above condition should be inverted. Instead of `err == errInvalidToken`, it should be `err != errInvalidToken`, because errInvalidToken is a known case, and we don't need to log it, but when we don't know what the error is, e.g. a network failure when talking to tokeninfo, we want to log what the error was.